sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1111 @@
+            KFold(self.n_folds, shuffle=self.shuffle,
+                  random_state=self._rng).split(np.zeros(max(c, self.n_folds)))
+            for c in label_counts]
+
+        test_folds = np.zeros(n_samples, dtype=np.int)
+        for test_fold_idx, per_label_splits in enumerate(zip(*per_label_cvs)):
+            for label, (_, test_split) in zip(unique_labels, per_label_splits):
+                label_test_folds = test_folds[y == label]
+                # the test split can be too big because we used
+                # KFold(...).split(X[:max(c, n_folds)]) when data is not 100%
+                # stratifiable for all the labels
+                # (we use a warning instead of raising an exception)
+                # If this is the case, let's trim it:
+                test_split = test_split[test_split < len(label_test_folds)]
+                label_test_folds[test_split] = test_fold_idx
+                test_folds[y == label] = label_test_folds
+
+        return test_folds
+
+    def _iter_test_masks(self, X, y=None, labels=None):
+        test_folds = self._make_test_folds(X, y, labels)
+        for i in range(self.n_folds):
+            yield test_folds == i
+
+
+class LeaveOneLabelOut(_PartitionIterator):
+    """Leave-One-Label_Out cross-validation iterator
+
+    Provides train/test indices to split data according to a third-party
+    provided label. This label information can be used to encode arbitrary
+    domain specific stratifications of the samples as integers.
+
+    For instance the labels could be the year of collection of the samples
+    and thus allow for cross-validation against time-based splits.
+
+    Read more in the :ref:`User Guide <split>`.
+
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import LeaveOneLabelOut
+    >>> X = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
+    >>> y = np.array([1, 2, 1, 2])
+    >>> labels = np.array([1, 1, 2, 2])
+    >>> lol = LeaveOneLabelOut()
+    >>> lol.n_splits(X, y, labels)
+    2
+    >>> print(lol)
+    LeaveOneLabelOut()
+    >>> for train_index, test_index in lol.split(X, y, labels):
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    ...    print(X_train, X_test, y_train, y_test)
+    TRAIN: [2 3] TEST: [0 1]
+    [[5 6]
+     [7 8]] [[1 2]
+     [3 4]] [1 2] [1 2]
+    TRAIN: [0 1] TEST: [2 3]
+    [[1 2]
+     [3 4]] [[5 6]
+     [7 8]] [1 2] [1 2]
+
+    """
+
+    def _iter_test_masks(self, X, y, labels):
+        # We make a copy of labels to avoid side-effects during iteration
+        labels = np.array(labels, copy=True)
+        unique_labels = np.unique(labels)
+        for i in unique_labels:
+            yield labels == i
+
+    def n_splits(self, X, y, labels):
+        return len(np.unique(labels))
+
+
+class LeavePLabelOut(_PartitionIterator):
+    """Leave-P-Label_Out cross-validation iterator
```

Comment:
Will fix this :) (btw the `_PartitionIterator` has been renamed to `_BaseCV` (Similar to `BaseEstimator`... does it sound good?)